Fix Jetpack Search pricing tier calculations

trakos opened the pull request:
#### Proposed Changes

Jetpack Search used pricing tiers data for some features.
However, with new pricing released recently, many of those
features stopped working.
New pricing uses quantity-based pricing, which requires
different type of calculations and more fields returned from API.
This PR requires diff D92804-code, which adds those fields.

List of fixes:

#### 1. Tooltip on https://cloud.jetpack.com/jetpack-search

Before:
![image](https://user-images.githubusercontent.com/6437642/202613679-8129eb3d-1827-4c42-84b0-c1f10a500503.png)

After:

![image](https://user-images.githubusercontent.com/6437642/202613743-8d3e5538-694b-4f93-88c2-d09e0c93615b.png)

#### 2. Quantity of requests/records being purchased for Jetpack Search during checkout

Before:

![image](https://user-images.githubusercontent.com/6437642/202613903-463c630d-f376-435e-a3ae-f2796d5e75ee.png)

After:

![image](https://user-images.githubusercontent.com/6437642/202613950-04cc3a75-c6f8-4230-851a-532124a6cf02.png)

Related discussion on P2: p7pQDF-7v2-p2#comment-21900
Related thread on twitter: https://twitter.com/redcrew/status/1593306122429005824

#### 3. Restored dropdown for switch between monthly and yearly billing

Before:

![image](https://user-images.githubusercontent.com/6437642/202614405-afeecfda-473e-4d76-a940-fff93b957854.png)

After:

![image](https://user-images.githubusercontent.com/6437642/202614351-dc652db5-4cae-4c46-b3e2-bd09d0e28300.png)


#### Testing Instructions

* Apply patch D92804-code on your sandbox
* Sandbox public-api.wordpress.com
* Start jetpack cloud 
* Visit http://jetpack.cloud.localhost:3000/jetpack-search and verify that the price tooltip shows correct amounts.
* Start calypso
* Visit http://calypso.localhost:3000/checkout/your_blog_url_here/jetpack_search and verify that the subtext under item line mentions how many records and/or requests will be purchased, and that it is possible to change to monthly pricing.


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

trakos committed:
Fix showing licensed quantity during checkout for Jetpack Search

Before we introduced new pricing for Jetpack Search,
there used to be a subtext under checkout line item
with tier name, for example "Up to 100 records".
This is not working with new pricing.

In this commit, I'm changing it so that we show text
like "Up to 20k records and/or requests per month"

trakos committed:
Fix tooltip with Jetpack Search tier explanations

On page https://cloud.jetpack.com/jetpack-search we have
a tooltip next to an estimated price explaining tiered pricing.
However, currently it shows prices of 0, as it hasn't been adjuster
to the new tiers systems.

Note that this requires changes introduced in diff D92804-code,
without it there's not enough data returned from API.

trakos committed:
Fix showing dropdown for monthly and yearly product variants for Jetpack Search

Before new pricing was introduced, there was an option
to choose between annual and monthly payments during checkout.
However, it broke with new pricing, as the price has to be
calculated using quantity-based tiered approach, instead of the
previous one, based on tiers alone.
Option to change variant disappeared after new pricing was introduced.

This diff restores that functionality by adding support for quantity-based
prices to `getSearchProductTierPrice` function.
It also adds a new parameter to `useGetProductVariants`, `current_quantity`.
Using posts count is not accurate, and pricing also has to depend on
count of requests in past months if user had a Jetpack Search subscription before.

Note that this requires changes introduced in diff D92804-code,
without it there's not enough data returned from API.

TheSteveK commented:
I'm testing with a JN site that has the Security plan.  I've got the diff applied to my sandbox for the API changes and public-api.wordpress.com pointed to the sandbox.  But, I'm seeing some strange results which I think the API changes were built to handle.  The pricing text is showing up based on 3 records, because I just have the default pages. So the tooltip shows "basded on 3 records" and the cart shows the pricing as up to 0.003k records per month:

<img width="552" alt="Screenshot 2022-11-22 at 9 09 19 AM" src="https://user-images.githubusercontent.com/12895386/203338007-c8ddf44a-e069-48c1-abed-e144aa0a7247.png">

This could be something weird in my environment, so another set of eyes would be appreciated :)

mattgawarecki commented:
@trakos My test results match the ones @TheSteveK reports:

- ✅ The Search page on Jetpack Cloud (and its accompanying tooltip) appears to be correct, showing me the right price tier information based on my record count.
- ⚠️ The site I tested with has 19 records, but I see "Up to 0.019k records and/or requests per month" at checkout (regardless of monthly/yearly billing term).
- ✅ The month/year dropdown is present at checkout, whereas it's not present in the current production environment.

<img width="450" alt="image" src="https://user-images.githubusercontent.com/670067/203422763-59da4631-8ecf-4f71-ad99-a64c75e67060.png">
<img width="489" alt="image" src="https://user-images.githubusercontent.com/670067/203422874-138ccb83-437a-45ac-a897-1cf39b23584b.png">

trakos committed:
Rename records to "records and/or requests per month"

trakos committed:
Fix search quantity displayed during checkout

trakos commented:
Thanks for the reviews @TheSteveK and @mattgawarecki!

My bad, I forgot that I mocked exactly 20k requests, and then assumed that the number is already rounded to the nearest 10k.

I've now updated D92804-code so that our cart endpoint returns tier quantity data as well, and fixed the amount shown in "Up to nk records and/or requests per month" text.

Additionally, I've updated the text `estimated price based on %(records)s records` to `estimated price based on %(records_and_or_requests)s records and/or monthly requests` on https://cloud.jetpack.com/jetpack-search. It's an important distinction when user considers upgrading from free tier, since we already have the number of monthly requests at this point and are using it when calculating the price.

mattgawarecki reviewed (approved):
After your most recents @trakos I went back to re-review this PR and everything seems to be correct now from a functional testing perspective. I gave the code itself a light read-through and didn't see anything objectionable, so I'm inclined to approve on behalf of Origin :shipit:

sirbrillig commented on packages/shopping-cart/src/types.ts:
```diff
@@ -507,6 +507,8 @@ export interface ResponseCartProduct {
 
 	price_tier_minimum_units?: number | null;
 	price_tier_maximum_units?: number | null;
+	price_tier_transform_quantity_divide_by?: number | null;
```
Could we put comments above each of these properties to explain what they mean?

trakos replied:
Good call, I've added comments to the added properties. I've mostly just copied existing ones from backend.

sirbrillig replied:
Lovely! 👌 Thank you!

sirbrillig reviewed (approved):
This looks good to me from a Payments perspective. Nice refactoring.

trakos committed:
Add comments to new price tier fields

trakos commented:
Thank you for reviews!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7636555).
